contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1047 @@
+    }
+    // slither-disable-next-line unused-return return value irrelevant
+    _addressToScheduleIdSet[recipient].add(projectId);
+    schedule.exists = true;
+    schedule.startTime = scheduleData.startTime;
+    schedule.endTime = scheduleData.startTime + restrictionDuration.duration;
+    schedule.releasedAmountFloor = 0;
+    emit ScheduleCreated(projectId, schedule.startTime, schedule.endTime);
+  }
+
+  /**
+   * @notice Revokes unreleased tokens from a schedule.
+   * This is an *admin* operation callable only by addresses having TOKEN_REVOKER_ROLE
+   * (enforced in `batchRevokeUnreleasedTokenAmounts`)
+   *
+   * @dev Only the maximum revocable number of tokens (unreleased tokens) can be revoked.
+   * Once the tokens have been revoked, the current released amount can never fall below
+   * its current level, even if the linear release schedule of the new amount would cause
+   * the released amount to be lowered at the current timestamp (a floor is established).
+   *
+   * Unlike in the `withdrawFromSchedule` function, here we burn `RestrictedNORI`
+   * from the schedule owner but send that `BridgedPolygonNORI` back to Nori's
+   * treasury or an address of Nori's choosing (the *to* address).
+   * The *claimedAmount* is not changed because this is not a claim operation.
+   */
+  function _revokeUnreleasedTokens(
+    address to,
+    uint256 removalId,
+    uint256 amount
+  ) internal {
+    // slither-disable-next-line calls-loop
+    uint256 scheduleId = _removal.getProjectIdForRemoval(removalId);
+    Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
+    if (!schedule.exists) {
+      revert NonexistentSchedule({scheduleId: scheduleId});
+    }
+    uint256 quantityToRevoke;
+    // amount of zero indicates revocation of all remaining tokens.
+    if (amount > 0) {
+      if (!(amount <= revocableQuantityForSchedule(scheduleId))) {
+        revert InsufficientUnreleasedTokens({scheduleId: scheduleId});
+      }
+      quantityToRevoke = amount;
+    } else {
+      quantityToRevoke = revocableQuantityForSchedule(scheduleId);
+    }
+
+    // burn correct proportion from each token holder
+    EnumerableSetUpgradeable.AddressSet storage tokenHolders = schedule
+      .tokenHolders;
+    uint256[] memory quantitiesToBurnForHolders = new uint256[](
+      tokenHolders.length()
+    );
+
+    uint256 cumulativeQuantityToBurn = 0;
+    for (uint256 i = 0; i < (tokenHolders.length() - 1); i++) {
+      uint256 quantityToBurnForHolder = _quantityToRevokePerTokenHolder(
+        quantityToRevoke,
+        scheduleId,
+        tokenHolders.at(i)
+      );
+      quantitiesToBurnForHolders[i] = quantityToBurnForHolder;
+      cumulativeQuantityToBurn += quantityToBurnForHolder;
+    }
+    quantitiesToBurnForHolders[tokenHolders.length() - 1] =
+      quantityToRevoke -
+      cumulativeQuantityToBurn;
+
+    for (uint256 i = 0; i < (tokenHolders.length()); i++) {
+      super._burn(
```

Comment:
should probably use burnBatch

Reply:
unfortunately we can't because `burnBatch` applies to a batch of token ids but _for a single address_ , whereas we are dealing with a single token id but multiple addresses in this case

Reply:
I see. Would be useful to write a todo here to consider writing a variant of burn batch that does that. Since this is a super call, it's not as important to batch it as if it were a call to another contract, but I think a batch variant of this would almost certainly reduce gas overhead, though I'd need to confirm that